glide-core/redis-rs/redis/src/cluster_async/pipeline_routing.rs

```diff
@@ -0,0 +1,519 @@
+/// - `cmd`: The original multi-slot command that needs to be split.
+/// - `index`: The index of the original command within the pipeline.
+/// - `slots`: A vector containing routing information. Each entry includes:
+///   - `Route`: The specific route for the slot.
+///   - `Vec<usize>`: Indices of the keys within the command that map to this slot.
+pub async fn handle_pipeline_multi_slot_routing<C>(
+    pipelines_by_connection: &mut NodePipelineMap<C>,
+    core: Core<C>,
+    cmd: &Cmd,
+    index: usize,
+    slots: Vec<(Route, Vec<usize>)>,
+) -> Result<(), (OperationTarget, RedisError)>
+where
+    C: Clone,
+{
+    // inner_index is used to keep track of the index of the sub-command inside cmd
+    for (inner_index, (route, indices)) in slots.iter().enumerate() {
+        let conn = {
+            let lock = core.conn_lock.read().expect(MUTEX_READ_ERR);
+            lock.connection_for_route(route)
+        };
+        if let Some((address, conn)) = conn {
+            // create the sub-command for the slot
+            let new_cmd = command_for_multi_slot_indices(cmd, indices.iter());
+            add_command_to_node_pipeline_map(
+                pipelines_by_connection,
+                address,
+                conn.await,
+                new_cmd,
+                index,
+                Some(inner_index),
+            );
+        } else {
+            return Err((
+                OperationTarget::NotFound,
+                RedisError::from((
+                    ErrorKind::ConnectionNotFoundForRoute,
+                    "No available connections",
+                )),
+            ));
+        }
+    }
+    Ok(())
+}
+
+/// Collects and sends pending requests for the given pipeline map, and waits for their responses.
+///
+/// This function creates `PendingRequest` objects for each pipeline in the provided pipeline map,
+/// adds them to the core's pending requests queue, and waits for all responses to be received.
+///
+/// # Arguments
+///
+/// * `pipeline_map` - A map of node pipelines where the commands are grouped by their corresponding nodes.
+/// * `core` - The core object that provides access to connection locks and other resources.
+///
+/// # Returns
+///
+/// A tuple containing:
+/// - A vector of results for each sub-pipeline execution.
+/// - A vector of (address, indices) pairs indicating where each response should be placed.
+pub async fn collect_and_send_pending_requests<C>(
+    pipeline_map: NodePipelineMap<C>,
+    core: Core<C>,
+) -> (
+    Vec<Result<RedisResult<Response>, RecvError>>,
+    Vec<(String, Vec<(usize, Option<usize>)>)>,
+)
+where
+    C: Clone + ConnectionLike + Connect + Send + Sync + 'static,
+{
+    // Processes the sub-pipelines to generate pending requests for execution on specific nodes.
+    // Each pending request encapsulates all the necessary details for executing commands on a node.
+    //
+    // Returns:
```

Comment:
you don't need to document the collect_pipeline_requests function here - you already do it in the function deceleration